lightning/src/routing/router.rs

```diff
@@ -1785,6 +1785,49 @@ fn add_random_cltv_offset(route: &mut Route, payment_params: &PaymentParameters,
 	}
 }
 
+/// Build a route from us (payer) with the given hops ending at the target node (payee).
+///
+/// Re-uses logic from `find_route`, so the restrictions described there also apply here.
+pub fn build_route_from_hops<L: Deref>(
+	our_node_pubkey: &PublicKey, hops: &[PublicKey], route_params: &RouteParameters, network: &NetworkGraph,
+	logger: L, random_seed_bytes: &[u8; 32]) -> Result<Route, LightningError>
+where L::Target: Logger {
+	let network_graph = network.read_only();
+	match build_route_from_hops_internal(
+		our_node_pubkey, hops, &route_params.payment_params, &network_graph,
+		route_params.final_value_msat, route_params.final_cltv_expiry_delta, logger, random_seed_bytes
+		) {
+		Ok(mut route) => {
+			add_random_cltv_offset(&mut route, &route_params.payment_params, &network_graph, random_seed_bytes);
+			Ok(route)
+		},
+		Err(err) => Err(err),
+	}
+}
+
+fn build_route_from_hops_internal<L: Deref>(
+	our_node_pubkey: &PublicKey, hops: &[PublicKey], payment_params: &PaymentParameters,
+	network_graph: &ReadOnlyNetworkGraph, final_value_msat: u64, final_cltv_expiry_delta: u32,
+	logger: L, random_seed_bytes: &[u8; 32]) -> Result<Route, LightningError> where L::Target: Logger {
+
+	let node_ids_iter = hops.iter().map(|hop_pubkey| NodeId::from_pubkey(&hop_pubkey));
+	let hop_iter = core::iter::once(NodeId::from_pubkey(&our_node_pubkey))
+		.chain(node_ids_iter.clone())
+		.zip(node_ids_iter);
+	let filter_channels = |_short_channel_id: u64, source: &NodeId, target: &NodeId,
+		_usage: ChannelUsage| {
```

Comment:
I would consider bringing all the parameters down to this line if they fit. I think rustfmt won't continue parameter lists in this way, FWIW. It will make sure all a parameters are at the same indentation.